python/pyarrow/types.pxi

```diff
@@ -2446,8 +2446,11 @@ cpdef DictionaryType dictionary(index_type, value_type, bint ordered=False):
         DictionaryType out = DictionaryType.__new__(DictionaryType)
         shared_ptr[CDataType] dict_type
 
-    if _index_type.id not in {Type_INT8, Type_INT16, Type_INT32, Type_INT64}:
-        raise TypeError("The dictionary index type should be signed integer.")
+    if _index_type.id not in {
+        Type_INT8, Type_INT16, Type_INT32, Type_INT64,
+        Type_UINT8, Type_UINT16, Type_UINT32, Type_UINT64,
```

Comment:
This set is actually pre-defined, but in the python wrapper:

https://github.com/apache/arrow/blob/6a6b4643611f34cac8c829e9dc41eb6e815abd7b/python/pyarrow/types.py#L28-L32

Instead of repeating it here, I could also move those set definitions to the cython file.

Reply:
Are these sets of types used in other files that if defined in types.pxi would prevent their visibility? If not, then I agree with moving them.

Reply:
There are related sets (such as [the primitive types](https://github.com/apache/arrow/blob/master/python/pyarrow/types.pxi#L1827)) already defined in `types.pxi`, so it makes sense to have them in the same file.

Reply:
Hmm, that primitive types definition in types.pxi actually doesn't seem to be used in practice .. 

(if we use more of those, it would certainly be worth moving them, though, I think. Something we can do when that comes up I suppose)